### DISCUSSION TITLE
Docker Workflow

### DISCUSSION THREAD

**Discussion by lpzhang-code** [community/community#26847] [Actions]
<p>Hello everyone!</p>
<p>I am new to GitHub Actions. I am wondering if there are experienced people who can offer advice on setting up CI pipeline involving Docker container.</p>
<p>Ideally, I would like to do the following inside the same job: (1) build the Docker image, (2) run unit tests inside the container, (3) push the built image up to Docker Hub if the tests pass.</p>

**Comment by airtower-luna** [comment]
<p>That’s definitely possible, I have an example of a similar setup here (pushing to GHCR instead): <a href="https://github.com/airtower-luna/hello-ghcr/blob/763f10f3fdc33e8bcef1d04508a3264cf1a06c16/.github/workflows/build.yaml">https://github.com/airtower-luna/hello-ghcr/blob/763f10f3fdc33e8bcef1d04508a3264cf1a06c16/.github/workflows/build.yaml</a></p>
<p>I just use the regular <code>docker</code> commands in the steps. Note that the <code>docker logout</code> at the end is run with <code>if: always()</code> to make sure the credentials are removed even if any intermediate step fails. There are also some pre-made actions like <a href="https://github.com/docker/login-action"><code>docker/login-action</code></a> available if you prefer.</p>
<p>Your tests will probably look pretty different, for the simple test container I only do <code>docker run</code> and check if the output looks as expected. :wink:</p>

**Comment by lpzhang-code** [comment]
<p>Thanks very much for this.</p>
<p>I don’t know much about writing logic inside of YAML files, do you have any suggested resources?</p>
<p>Specifically, I want to write logic where I have this step</p>
<pre><code class="lang-auto">- name: Build Docker Image
   run: docker build -t image_name:TAG_NAME .

</code></pre>
<p>where TAG_NAME is ‘staging’ if I push to develop branch and ‘prod’ if I push to master branch</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="lpzhang5" data-post="3" data-topic="134125">
<div class="title">
<div class="quote-controls"></div>
 lpzhang5:</div>
<blockquote>
<p>I don’t know much about writing logic inside of YAML files, do you have any suggested resources?</p>
</blockquote>
</aside>
<p>There isn’t much. YAML itself is structured data. Github Actions defines some fields that can contain logic (mostly <code>if:</code> on steps or jobs). Most of the logic in your workflow will have to be in the steps.</p>
<aside class="quote no-group" data-username="lpzhang5" data-post="3" data-topic="134125">
<div class="title">
<div class="quote-controls"></div>
 lpzhang5:</div>
<blockquote>
<p>where TAG_NAME is ‘staging’ if I push to develop branch and ‘prod’ if I push to master branch</p>
</blockquote>
</aside>
<p>You might have noticed this step in my example:</p>
<pre><code class="lang-yaml">      - name: Generate container tag(s) and name
        run: |
          TAGS=$(python3 tag-from-ref.py ${{ github.ref }})
          echo "::set-env name=TAGS::${TAGS}"
          echo "::set-env name=CONTAINER_NAME::${{ github.repository_owner }}/hello-ghcr-meow"
</code></pre>
<p>The <a href="https://github.com/airtower-luna/hello-ghcr/blob/763f10f3fdc33e8bcef1d04508a3264cf1a06c16/tag-from-ref.py"><code>tag-from-ref.py</code></a> processes <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context"><code>${{ github.ref }}</code></a> to decide which tag(s) to use, and the result is then stored in an environment variable.</p>
<p>Whether it makes sense to create a script file like I did or put the script directly into a step is part a matter of taste and also depends on how complex your selection is. You could just do something like this for what you described:</p>
<pre><code class="lang-yaml">      - name: Generate container tag
        run: |
          if [ "${{ github.ref }}" = "refs/heads/develop" ]; then
              echo "::set-env name=TAG_NAME::staging"
          elif [ "${{ github.ref }}" = "refs/heads/master" ]; then
              echo "::set-env name=TAG_NAME::prod"
          else
              # whatever you want to do for other branches/tags
          fi
</code></pre>

**Comment by lpzhang-code** [comment]
<p>Thanks for your help.</p>
<p>I have adapted the logic you have written and workflow is working as intended.</p>
<p>I have also learned quite a bit about the general process, previously I was very unfamiliar.</p>
